lib/THC/THCReduce.cuh

```diff
@@ -200,11 +223,33 @@ bool THC_reduceDim(THCState* state,
     block = getContigReduceBlock(outElements, reductionSize);
     smemSize = sizeof(typename TensorUtils<TensorType>::DataType) * block.x;
   } else {
-    if (!getNoncontigReduceGrid(outElements, grid)) {
-      return false;
+
+    //If there are a large number of outputs to the reduction, avoid syncthreads
+    //in kernel
+    if(outElements > 512*16){
+      block = dim3(512);
+    }else{
+      //x dim does different slices
+      //y dim helps with a slice
+      //If we only have 8 loops, don't bother sharing work across ydim
+      unsigned long ydim = THCCeilDiv(reductionSize, 16L);
+
+      ydim = min((unsigned long) 16, ydim);
+
+      block = dim3(512, 1, 1);
+      while(ydim > 1){
+        block.x /= 2;
+        block.y *= 2;
+        ydim /= 2;
+      }
+    }
+    long gridx  = THCCeilDiv( outElements, (long)block.x);
+    if (gridx > 1024){
+      long n_loops = THCCeilDiv(outElements, (long) (1024 * block.x) );
+      gridx = outElements / (block.x*n_loops);
```

Comment:
Are you sure this is ok? If you remove the ceil it is equivalent to setting gridx to 1024.